### PR TITLE
support ScatterND(18) and ScatterElement(18)

### PR DESCRIPTION
**Pull request by liqunfu**
Signed-off-by: Liqun Fu <liqfu@microsoft.com>

### Description
support ScatterND(18) and ScatterElement(18)

### Motivation and Context
This is for Ort 1.14 release

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc [hunk 1]
```diff
@@ -263,6 +272,84 @@ struct Func_Mul_ND<BFloat16> {
...
+      a++;
+      b++;
+    }
+  }
+};
+
+template <>
+struct Func_Max_ND<bool> {
+  void operator()(bool*, const bool*, uint64_t) const {
+    ORT_NOT_IMPLEMENTED("CPU execution provider: bool data type is not supported with ScatterND opset 18 when reduction is 'max'.");
```
Comment (BowenBao) [inline onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:328]: For my own education. Why are these types unsupported for "Min/Max"?

File: onnxruntime/test/testdata/onnx_backend_test_series_filters.jsonc [hunk 2]
```diff
@@ -303,38 +300,41 @@
...
-        "^test_sequence_insert_at_front",
-        "^test_maxpool_with_argmax_2d_precomputed_strides", //Disabled as it throws segfault
-        "^test_maxpool_with_argmax_2d_precomputed_pads", //Disabled as it throws segfault
-        "^test_resize_downsample_scales_cubic", // Runs but there's accuracy mismatch
-        "^test_resize_downsample_scales_linear", // Runs but there's accuracy mismatch
-        "^test_training_dropout_default_mask", // Runs but there's accuracy mismatch
-        "^test_training_dropout_mask", // Runs but there's accuracy mismatch
-        "^test_training_dropout_default" // Runs but there's accuracy mismatch
-    ],
+  "current_failing_tests_OPENVINO_CPU_FP32": [
```
Comment (xadupre) [inline onnxruntime/test/testdata/onnx_backend_test_series_filters.jsonc:303]: I don't see any change in this section except the indentation.

**Comment by edgchen1** [comment]
should we add unit tests to test the new functionality?

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 1 above)
Comment (liqunfu) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:328]: I think Min/Max only apply to numeric data types. It does not make sense, for example, to do Max(true, false).

**Comment by liqunfu** [comment]
> should we add unit tests to test the new functionality?

In general, ONNX test cases are tested with ORT. I do added test in some PRs aimed at release 1.14. I may add test cases in this PR if time allows for the branch cut.

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 1 above)
Comment (gramalingam) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:328]: Hmmm. I guess it is a question for the ONNX spec to clarify. If this is the case, the ONNX spec should be updated. Min/Max is ambiguous for bool: we could interpret false/true as 0/1 and give it a meaning.

**Review by gramalingam — APPROVED** [review]
LGTM, except for minor nits.

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 1 above)
Comment (BowenBao) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:328]: And why is `MLFloat16` and `BFloat16` skipped?

And regarding Min/Max for bool, the convention seems to interpret it as 0/1 as Rama mentioned.
```python
>>> max(True, False)
True
>>> import torch
>>> torch.max(torch.tensor([True, False]))
tensor(True)
>>> import numpy
>>> numpy.max(numpy.array([True, False]))
True
```

Comment (liqunfu) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:328]: 16 float is ingeneral not supported with cpu provider.
re false/true to 0/1, I think it is better to add OR/And/Xor logical reduction mothod

File: onnxruntime/test/testdata/onnx_backend_test_series_filters.jsonc (hunk 2 above)
Comment (liqunfu) [reply onnxruntime/test/testdata/onnx_backend_test_series_filters.jsonc]: good catch! reverted the change

**Review by gramalingam — APPROVED** [review]

**Pull request merged**